.github/setup-node/action.yml

```diff
@@ -0,0 +1,50 @@
+name: 'Setup Node'
+description: 'Setup node and install dependencies with caching support'
+inputs:
+    node-version:
+        description: 'Optional node version to pass to setup-node'
+        required: false
+        type: string
+
+runs:
+    using: 'composite'
+    steps:
+        - name: Use desired version of NodeJS from input
+          if: ${{ inputs.node-version != '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version: ${{ inputs.node-version }}
+              cache: npm
+
+        - name: Use desired version of NodeJS from .nvmrc
+          if: ${{ inputs.node-version == '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version-file: '.nvmrc'
+              cache: npm
+
+        - name: Get node and npm version
+          id: node-npm-version
+          run: |
+              echo "NODE_VERSION=$(node -v)" >> $GITHUB_OUTPUT
+              echo "NPM_VERSION=$(npm -v)" >> $GITHUB_OUTPUT
+          shell: bash
+
+        - name: Cache node_modules
+          id: cache-node_modules
+          uses: actions/cache@v3
+          with:
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-npm-version.outputs.NODE_VERSION }}/${{ steps.node-npm-version.outputs.NPM_VERSION }}-${{ hashFiles('package-lock.json') }}
```

Comment:
doesn't this happen regularly in Gutenberg since our `npm` version is still gated at `<7`?

Reply:
I don't think we need to include `npm` version here too. Including it will potentially bust the cache more often.

> doesn't this happen regularly in Gutenberg since our `npm` version is still gated at `<7`?

That's because we're using Node 14, which comes with npm v6 by default. npm versions could change independently from Node.js, but they should just be minor releases and backward-compatible.

Reply:
Right, but according to the project's `package.json` file we're good to use `node@latest`, which is I think what most people end up doing locally.

https://github.com/WordPress/gutenberg/blob/6434e0bc181bb1c0f5fcf48520be20f1a37afb6c/package.json#L17-L20

so maybe by happenstance we're only using `node@14` in the test suites, but even then, are we guaranteed to run the specific version of 14? or are we pinning it entirely at a specific version?

Reply:
> so maybe by happenstance we're only using node@14 in the test suites

`.nvmrc` pins the Node.js version to 14 on CI. FWIW, there's an ongoing PR (#48950) that hopefully can be merged soon to upgrade Node.js and npm to latest on CI.

> but even then, are we guaranteed to run the specific version of 14

Nope, if there's a minor/patch release then we'll automatically use the newer version on CI (basically `^14`). But that's okay since we include the full node version in the cache key.

Note that this PR doesn't affect anyone's local setup whatsoever, they can continue using whatever supported Node.js/npm version they like locally. And the cache only exists on CI, so they won't be affected by it either. This PR only caches `node_modules` on CI, which is a controlled environment.

Reply:
> if there's a minor/patch release then we'll automatically use the newer version on CI (basically ^14). But that's okay since we include the full node version in the cache key.

I think we're in agreement here. my comment was specifically commenting on how we _do_ expect changes to `node` to occur that don't match `npm`. we're intentionally not using the bundled version of `npm` in the project, but we build the project with changing versions of `node` quite often, since we pin `node` at `>=` a major version.